lib/concurrency/Prefix_Refinement.thy

```diff
@@ -106,74 +95,218 @@ lemmas is_matching_fragment_env_closed = is_matching_fragment_wf[THEN conjunct2,
 lemmas split_iffD1 = Product_Type.split[THEN iffD1]
 lemmas pfx_refnD = prefix_refinementD
 lemmas pfx_refnD2 = pfx_refnD[THEN split_iffD1[where a=tr and b=res for tr res], rule_format]
 
+lemma prefix_refinement_False:
+  "prefix_refinement sr isr osr rvr AR R P \<bottom>\<bottom> f g"
+  by (clarsimp simp: prefix_refinement_def)
+
+lemma prefix_refinement_False':
+  "prefix_refinement sr isr osr rvr AR R \<bottom>\<bottom> Q f g"
+  by (clarsimp simp: prefix_refinement_def)
+
 lemma matching_tr_pfx_aCons:
   "matching_tr_pfx sr ((tmid, s) # atr) ctr
     = (\<exists>s'. length atr < length ctr \<and> rev ctr ! length atr = (tmid, s')
-        \<and> sr s s' \<and> matching_tr_pfx sr atr ctr)"
+            \<and> sr s s' \<and> matching_tr_pfx sr atr ctr)"
   apply (simp add: matching_tr_pfx_def matching_tr_def Suc_le_eq
                    list_all2_conv_all_nth less_Suc_eq all_conj_distrib)
   apply (simp add: nth_append prod_eq_iff)
   apply auto
   done
 
 lemma rely_cond_hd:
-  "rely_cond R s0 xs \<Longrightarrow> xs \<noteq> []
-    \<Longrightarrow> fst (hd xs) = Env \<longrightarrow> R (last_st_tr (tl xs) s0) (snd (hd xs))"
+  "\<lbrakk>rely_cond R s0 xs; xs \<noteq> []\<rbrakk>
+   \<Longrightarrow> fst (hd xs) = Env \<longrightarrow> R (last_st_tr (tl xs) s0) (snd (hd xs))"
   by (clarsimp simp: rely_cond_def neq_Nil_conv trace_steps_append
               split: if_split_asm)
 
-lemma diff_Suc_eq_if:
-  "(Suc n - m) = (if m \<le> n then Suc (n - m) else 0)"
-  by auto
-
 lemma rely_cond_nth:
-  "rely_cond R s0 tr \<Longrightarrow> n < length tr
-    \<Longrightarrow> fst (rev tr ! n) = Env \<longrightarrow> R ((if n = 0 then s0 else snd (rev tr ! (n - 1)))) (snd (rev tr ! n))"
+  "\<lbrakk>rely_cond R s0 tr; n < length tr\<rbrakk>
+   \<Longrightarrow> fst (rev tr ! n) = Env \<longrightarrow> R ((if n = 0 then s0 else snd (rev tr ! (n - 1)))) (snd (rev tr ! n))"
   by (simp add: rely_cond_def trace_steps_rev_drop_nth[where n=0, simplified])
 
 lemma is_matching_fragment_Nil:
-  "is_matching_fragment sr osr rvr ctr cres s0 R s f
-    \<Longrightarrow> [] \<in> fst ` f s"
+  "is_matching_fragment sr osr rvr ctr cres s0 R s f \<Longrightarrow> [] \<in> fst ` f s"
   apply (clarsimp simp: is_matching_fragment_def)
   apply (clarsimp simp only: set_eq_iff empty_iff simp_thms not_all)
   apply (drule(1) prefix_closed_drop[where tr=tr and n="length tr" for tr])
-  apply (clarsimp simp add: in_fst_snd_image)
+  apply (clarsimp simp: in_fst_snd_image)
+  done
+
+\<comment> \<open>FIXME: it would be good to show this but the last assumption needs to be a lot more complicated
+          and the nondet equivalent isn't used anywhere.
+lemma prefix_refinement_propagate_no_fail:
+  "\<lbrakk>prefix_refinement sr isr osr rvr AR R P Q f f';
+    no_fail P' f; \<forall>t0 t. Q t0 t \<longrightarrow> (\<exists>s0 s. P s0 s \<and> sr s0 t0 \<and> isr s t)\<rbrakk>
+   \<Longrightarrow> no_fail Q' f'"
```

Comment:
It's interesting that this doesn't fall out easily. It's one of the fundamental properties of refinement. We don't make explicit use of it, but it's a good sanity check and it could be a sign that our definition of data refinement is still a bit too complex. No need to resolve this in the short term, but maybe there is a tweak or two yet to be discovered that might make things easier in the future. (It's not unexpected that there would be side conditions since the type for programs is a lot tighter in the nondet monad, whereas for the trace monad there are many traces that don't really make sense as programs. So maybe it's just a reflection of that).

Reply:
This was one of the things I tried to prove pretty early on in the process of working on all this, so it might be easier now that I have a better understanding of the `prefix_refinement` definition. I was sort of thinking to leave it though, until we have a chance to put more thought into how we want to handle failure. Currently it is very different to how `corres` handles it.

Reply:
Having played around with this a bit more, it's actually a bit more complicated than I was previously thinking. The current definition allows us to show refinement between a non-failing abstract function and a failing concrete function, as long as all of the failing traces in the concrete function do not satisfy the rely.

It is almost definitely possible to resolve this, it will just require a bit of thought and possible some worked examples to find the nicest way off expressing things.

Reply:
No need to do this now. As long as it's flagged we can do this at any time later.